embassy-time/src/queue.rs

```diff
@@ -0,0 +1,617 @@
+        self.dispatch(alarm_schedule);
+    }
+
+    fn dispatch(&mut self, alarm_schedule: &AtomicU64) {
+        let now = Instant::now();
+
+        while self.queue.peek().filter(|timer| timer.at <= now).is_some() {
+            self.queue.pop().unwrap().waker.wake();
+        }
+
+        self.update_alarm(alarm_schedule);
+    }
+
+    fn update_alarm(&mut self, alarm_schedule: &AtomicU64) {
+        if let Some(timer) = self.queue.peek() {
+            let new_at = timer.at;
+
+            if self.alarm_at != new_at {
+                self.alarm_at = new_at;
+                alarm_schedule.store(new_at.as_ticks(), AtomicOrdering::SeqCst);
+            }
+        } else {
+            self.alarm_at = Instant::MAX;
+            alarm_schedule.store(Instant::MAX.as_ticks(), AtomicOrdering::SeqCst);
+        }
+    }
+
+    fn handle_alarm(&mut self, alarm_schedule: &AtomicU64) {
+        self.alarm_at = Instant::MAX;
+
+        self.dispatch(alarm_schedule);
+    }
+}
+
+/// The generic queue implementation
+pub struct Queue<const N: usize = 128, R: RawMutex = CriticalSectionRawMutex> {
+    inner: Mutex<R, RefCell<InnerQueue<N>>>,
+    alarm: Cell<Option<AlarmHandle>>,
+    alarm_schedule: AtomicU64,
+}
+
+impl<const N: usize, R: RawMutex + 'static> Queue<N, R> {
+    /// Create a Queue
+    pub const fn new() -> Self {
+        Self {
+            inner: Mutex::new(RefCell::new(InnerQueue::<N>::new())),
+            alarm: Cell::new(None),
+            alarm_schedule: AtomicU64::new(u64::MAX),
+        }
+    }
+
+    /// Initialize the queue
+    ///
+    /// This method is called from [`initialize`](crate::queue::initialize), so you are not expected to call it directly.
+    /// Call [`initialize`](crate::queue::initialize) instead.
+    ///
+    /// # Safety
+    /// It is UB call this function more than once, or to call it after any of your
+    /// futures that use `embassy-time` are polled already.
+    pub unsafe fn initialize(&'static self) {
+        if self.alarm.get().is_some() {
+            panic!("Queue is already initialized");
+        }
+
+        let handle = allocate_alarm().unwrap();
+        self.alarm.set(Some(handle));
+
+        set_alarm_callback(handle, Self::handle_alarm_callback, self as *const _ as _);
+    }
+
+    /// Schedule a new waker to be awoken at moment `at`
+    ///
+    /// This method is called internally by [`embassy-time`](crate), so you are not expected to call it directly.
+    pub fn schedule(&'static self, at: Instant, waker: &Waker) {
+        self.check_initialized();
+
+        self.inner
+            .lock(|inner| inner.borrow_mut().schedule(at, waker, &self.alarm_schedule));
+
+        self.update_alarm();
+    }
+
+    fn check_initialized(&self) {
+        if self.alarm.get().is_none() {
+            panic!("Queue is not initialized yet");
+        }
+    }
+
+    fn update_alarm(&self) {
+        // Need to set the alarm when we are *not* holding the mutex on the inner queue
```

Comment:
I'm starting to think we should forbid `set_alarm` from calling the callback synchronously. It could return an "instant already in the past" error instead, so the timer queue knows it's running behind and it does another "round" of expirations. 

It'd save a lot of trouble here, because it'd allow putting all the state in `InnerQueue`. It wouldn't make the executor-tied impl much more complex either, it'd just mean doing another executor loop on set_alarm error.

Reply:
+1

Reply:
This is addressed in the latest version of the PR, where the contract of `Driver::set_alarm` is changed.